Src/DependencyCollector/Shared/ServiceBusDiagnosticsListener.cs

```diff
@@ -0,0 +1,172 @@
+        // I.e. this dictionary is quite small, does not grow up after service warm-up and does not require clean up
+        private readonly ConcurrentDictionary<Property, PropertyFetcher> propertyFetchers = new ConcurrentDictionary<Property, PropertyFetcher>();
+
+        internal ServiceBusDiagnosticsEventHandler(TelemetryConfiguration configuration)
+        {
+            this.telemetryClient = new TelemetryClient(configuration);
+        }
+
+        /// <summary>
+        /// Gets custom handler <see cref="HandleDiagnosticsEvent{TContext}"/>delegate.
+        /// </summary>
+        internal HandleDiagnosticsEvent<HashSet<string>> EventHandler => this.HandleEvent;
+
+        private void HandleEvent(KeyValuePair<string, object> evnt, DiagnosticListener diagnosticListener, HashSet<string> context)
+        {
+            // there is no guarantee IsEnabled was called in case there are multiple listeners in the system. i.e. we need to check it again
+            if (evnt.Key.EndsWith("Start", StringComparison.Ordinal))
+            {
+                return;
+            }
+
+            Activity currentActivity = Activity.Current;
+
+            switch (evnt.Key)
+            {
+                case "Microsoft.Azure.ServiceBus.ProcessSession.Stop":
+                case "Microsoft.Azure.ServiceBus.Process.Stop":
+                    this.OnRequest(evnt.Key, evnt.Value, currentActivity);
+                    break;
+                case "Microsoft.Azure.ServiceBus.Exception":
+                    this.OnException(evnt.Key, evnt.Value);
+                    break;
+                default:
+                    this.OnDependency(evnt.Key, evnt.Value, currentActivity);
+                    break;
+            }
+        }
+
+        private void OnDependency(string name, object payload, Activity activity)
+        {
+            DependencyTelemetry telemetry = new DependencyTelemetry { Type = DependencyType };
+
+            this.SetCommonProperties(name, payload, activity, telemetry);
+
+            // Endpoint is URL of particular ServiceBus, e.g. sb://myservicebus.servicebus.windows.net/
+            telemetry.Data = this.FetchPayloadProperty<Uri>(name, EndpointPropertyName, payload)?.ToString();
+
+            // Queue/Topic name, e.g. myqueue/mytopic
+            telemetry.Target = this.FetchPayloadProperty<string>(name, EntityPropertyName, payload);
+
+            this.telemetryClient.TrackDependency(telemetry);
+        }
+
+        private void OnRequest(string name, object payload, Activity activity)
+        {
+            RequestTelemetry telemetry = new RequestTelemetry();
+
+            this.SetCommonProperties(name, payload, activity, telemetry);
+
+            // Endpoint is URL of particular ServiceBus, e.g. sb://myservicebus.servicebus.windows.net/
+            telemetry.Url = this.FetchPayloadProperty<Uri>(name, EndpointPropertyName, payload);
```

Comment:
So request URL traditionally identifies something on the processing side (server) - here the Endpoint identifies the triggering side (client). I know Endpoint is the only URL-like data that we have but I think it might be confusing. We might just leave it blank

Reply:
Whatever we set in the dependency.Data we should set in request.Url.
As I shown above it could be useful to distinguish different SBs based on an endpoint and in this case, it's better to have it in the properties rather than in custom dimensions.

Reply:
fair enough - Data and Url should match, still not sure what is the value of putting Endpoint here as it is the same for a single SB
the picture that you're painting is that Entity (queue name) is a higher level thing than the actual SB (Endpoint)

Reply:
it's the most granular and verbose thing we have: it has the most details. You may have 10 queues in one servicebus. Or in case of Azure Storage Queues, queuename is all you have (endpoint is always the same and different for emulators/private/govenrments/etc clouds only).
 We are trying here not only for service bus, we are trying to make sure the same work for other queues as well.